src/cs50/sql.py

```diff
@@ -90,9 +95,10 @@ def connect(dbapi_connection, connection_record):
             self._logger.disabled = disabled
 
     def __del__(self):
-        """Close database connection."""
-        if hasattr(self, "_connection"):
-            self._connection.close()
+        """Close database session and connection."""
+        if self._session is not None:
+            self._session.close()
+            self._session = None
```

Comment:
Instead of repeatedly using this condition and setting the session to `None` all over the code, should we perhaps factor this out into a `_close_session` method? For example:

```
def _close_session(self):
  if self._session is not None:
    self._session.close()

  self._session = None
```

That way we could just call `self._close_session()` then optionally instantiate a new session using `self._session = self._Session(...)`.

Reply:
We could also add a `_get_session` method that either returns the current session or instantiates a new session depending on what we want. For example:

```
def _get_session(create=True, args=[], kwargs={}):
  if create and self._session is None:
    self._session = self._Session(*args, **kwargs)

  return self._session
```

That way, instead of doing something like:

```
if self._session is None:
  self._session = self._Session()
```

we'd just do:

```
session = self._get_session()
```

Although, I might be slightly over engineering this given that we're instantiating a session in only two places now.

Reply:
Yeah, and I'm not sure I can think of any other times in execute where we'd need to start a transaction -- right now, the two places are "the library user is starting a transaction explicitly with BEGIN" and "we are starting a transaction for the sake of autocommit." Which pretty much covers the range of when a transaction should start I think?

Reply:
@kzidane I'm adding `_close_session` after `execute` and before `_escape` because that seems most logical in terms of code organization (special methods, public functions, private functions)? But please do advise if my understanding is wrong.

Reply:
Updated the code; we're down to one session creation per call to `execute`, so I'm going to resolve this conversation.